File: src/main/java/org/jukeboxmc/math/Vector.java
package org.jukeboxmc.math;

import com.nukkitx.math.vector.Vector3d;
import com.nukkitx.math.vector.Vector3f;
import com.nukkitx.math.vector.Vector3i;
import com.nukkitx.math.vector.Vector3l;
import lombok.SneakyThrows;
import org.apache.commons.math3.util.FastMath;
import org.jukeboxmc.world.Dimension;

import java.util.Objects;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class Vector implements Cloneable {

    protected float x;
    protected float y;
    protected float z;
    protected Dimension dimension = Dimension.OVERWORLD;

    public Vector( float x, float y, float z, Dimension dimension ) {
        this.x = x;
        this.y = y;
        this.z = z;
        this.dimension = dimension;
    }

    public Vector( int x, int y, int z, Dimension dimension ) {
        this.x = x;
        this.y = y;
        this.z = z;
        this.dimension = dimension;
    }

    public Vector( float x, float y, float z ) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public Vector( int x, int y, int z ) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public Vector( Vector3f vector3f ) {
        this.x = vector3f.getX();
        this.y = vector3f.getY();
        this.z = vector3f.getZ();
    }

    public Vector( Vector3i vector3i ) {
        this.x = vector3i.getX();
        this.y = vector3i.getY();
        this.z = vector3i.getZ();
    }

    public static Vector zero() {
        return new Vector( 0, 0, 0 );
    }

    public static Vector up() {
        return new Vector( 0, 1, 0 );
    }

    public static Vector down() {
        return new Vector( 0, -1, 0 );
    }

    public static Vector north() {
        return new Vector( 0, 0, -1 );
    }

    public static Vector east() {
        return new Vector( 1, 0, 0 );
    }

    public static Vector south() {
        return new Vector( 0, 0, 1 );
    }

    public static Vector west() {
        return new Vector( -1, 0, 0 );
    }

    public float getX() {
        return this.x;
    }

    public float getY() {
        return this.y;
    }

    public float getZ() {
        return this.z;
    }

    public void setX( final float x ) {
        this.x = x;
    }

    public void setY( final float y ) {
        this.y = y;
    }

    public void setZ( final float z ) {
        this.z = z;
    }

    public int getBlockX() {
        return (int) FastMath.floor( this.x );
    }

    public int getBlockY() {
        return (int) FastMath.floor( this.y );
    }

    public int getBlockZ() {
        return (int) FastMath.floor( this.z );
    }

    public int getChunkX() {
        return this.getBlockX() >> 4;
    }

    public int getChunkZ() {
        return this.getBlockZ() >> 4;
    }
    
    public Dimension checkDimensionIsNull() { // i've added this to fix <Cannot invoke "org.jukeboxmc.math.Vector.getDimension()" because "vector" is null> error
        return this.dimension;
    }
    
    public Dimension getDimension() {
        return (Vector.checkDimensionIsNull() == null) ? 0 : this.dimension;
    }

    public void setDimension( Dimension dimension ) {
        this.dimension = dimension;
    }

    public Vector getVectorWhenXIsOnLine( Vector other, float x ) {
        float xDiff = other.x - this.x;
        float yDiff = other.y - this.y;
        float zDiff = other.z - this.z;

        float f = ( x - this.x ) / xDiff;
        return ( f >= 0F && f <= 1F ) ? new Vector( this.x + xDiff * f, this.y + yDiff * f, this.z + zDiff * f, this.dimension ) : null;
    }

    public Vector getVectorWhenYIsOnLine( Vector other, float y ) {
        float xDiff = other.x - this.x;
        float yDiff = other.y - this.y;
        float zDiff = other.z - this.z;

        float f = ( y - this.y ) / yDiff;
        return ( f >= 0F && f <= 1F ) ? new Vector( this.x + xDiff * f, this.y + yDiff * f, this.z + zDiff * f, this.dimension ) : null;
    }

    public Vector getVectorWhenZIsOnLine( Vector other, float z ) {
        float xDiff = other.x - this.x;
        float yDiff = other.y - this.y;
        float zDiff = other.z - this.z;

        float f = ( z - this.z ) / zDiff;
        return ( f >= 0F && f <= 1F ) ? new Vector( this.x + xDiff * f, this.y + yDiff * f, this.z + zDiff * f, this.dimension ) : null;
    }

    public Vector add( final float x, final float y, final float z ) {
        return new Vector( this.x + x, this.y + y, this.z + z, this.dimension );
    }

    public Vector subtract( final float x, final float y, final float z ) {
        return new Vector( this.x - x, this.y - y, this.z - z, this.dimension );
    }

    public Vector multiply( final float x, final float y, final float z ) {
        return new Vector( this.x * x, this.y * y, this.z * z, this.dimension );
    }

    public Vector divide( final float x, final float y, final float z ) {
        return new Vector( this.x / x, this.y / y, this.z / z, this.dimension );
    }

    public Vector normalize() {
        final float squaredLength = this.squaredLength();
        return this.divide( squaredLength, squaredLength, squaredLength );
    }

    public float distance( final Vector vector ) {
        return (float) Math.sqrt( this.distanceSquared( vector ) );
    }

    public float distanceSquared( final Vector vector ) {
        return (float) ( FastMath.pow( ( this.x - vector.getX() ), 2 ) + FastMath.pow( ( this.y - vector.getY() ), 2 ) +
                FastMath.pow( ( this.z - vector.getZ() ), 2 ) );
    }

    public float squaredLength() {
        return (float) ( FastMath.sqrt( this.x * this.x + this.y * this.y + this.z * this.z ) );
    }

    public Vector3f toVector3f() {
        return Vector3f.from( this.x, this.y, this.z );
    }

    public Vector3d toVector3D() {
        return Vector3d.from( this.x, this.y, this.z );
    }

    public Vector3l toVector3l() {
        return Vector3l.from( this.x, this.y, this.z );
    }

    public Vector3i toVector3i() {
        return Vector3i.from( this.x, this.y, this.z );
    }

    @Override
    public boolean equals( Object o ) {
        if ( this == o ) return true;
        if ( o == null || getClass() != o.getClass() ) return false;
        Vector vector = (Vector) o;
        return Float.compare( vector.x, this.x ) == 0 && Float.compare( vector.y, this.y ) == 0 && Float.compare( vector.z, this.z ) == 0 && this.dimension == vector.dimension;
    }

    @Override
    public int hashCode() {
        return Objects.hash( this.x, this.y, this.z, this.dimension );
    }

    @SneakyThrows
    @Override
    public Vector clone(){
        Vector clone = (Vector) super.clone();
        clone.x = this.x;
        clone.y = this.y;
        clone.z = this.z;
        return clone;
    }

    @Override
    public String toString() {
        return "Vector{" +
                "x=" + x +
                ", y=" + y +
                ", z=" + z +
                ", dimension=" + dimension +
                '}';
    }
}

File: JukeboxMC-API/src/main/kotlin/org/jukeboxmc/api/math/Vector.kt
package org.jukeboxmc.api.math

import lombok.SneakyThrows
import org.apache.commons.math3.util.FastMath
import org.jukeboxmc.api.world.Dimension
import java.util.*
import kotlin.math.roundToInt
import kotlin.math.sqrt


open class Vector : Cloneable {

    private var x: Float = 0F
    private var y: Float = 0F
    private var z: Float = 0F
    private var dimension: Dimension = Dimension.OVERWORLD

    constructor()

    constructor(x: Float, y: Float, z: Float, dimension: Dimension) {
        this.x = x
        this.y = y
        this.z = z
        this.dimension = dimension
    }

    constructor(x: Int, y: Int, z: Int, dimension: Dimension) {
        this.x = x.toFloat()
        this.y = y.toFloat()
        this.z = z.toFloat()
        this.dimension = dimension
    }

    constructor(x: Float, y: Float, z: Float) {
        this.x = x
        this.y = y
        this.z = z
    }

    constructor(x: Int, y: Int, z: Int) {
        this.x = x.toFloat()
        this.y = y.toFloat()
        this.z = z.toFloat()
    }

    fun up(): Vector {
        return Vector(x, y + 1, z)
    }

    fun down(): Vector {
        return Vector(x, y - 1, z)
    }

    fun north(): Vector {
        return Vector(x, y, z - 1)
    }

    fun east(): Vector {
        return Vector(x + 1, y, z)
    }

    fun south(): Vector {
        return Vector(x, y, z + 1)
    }

    fun west(): Vector {
        return Vector(x - 1, y, z)
    }

    fun setVector(x: Int, y: Int, z: Int) {
        this.x = x.toFloat()
        this.y = y.toFloat()
        this.z = z.toFloat()
    }

    fun getX(): Float {
        return this.x
    }

    fun setX(x: Float) {
        this.x = x
    }

    fun getY(): Float {
        return this.y
    }

    fun setY(y: Float) {
        this.y = y
    }

    fun getZ(): Float {
        return this.z
    }

    fun setZ(z: Float) {
        this.z = z
    }

    fun getBlockX(): Int {
        return FastMath.floor(x.toDouble()).toInt()
    }

    fun getBlockY(): Int {
        return FastMath.floor(y.toDouble()).toInt()
    }

    fun getBlockZ(): Int {
        return FastMath.floor(z.toDouble()).toInt()
    }

    fun getDimension(): Dimension {
        return this.dimension
    }

    fun getChunkX(): Int {
        return getBlockX() shr 4
    }

    fun getChunkZ(): Int {
        return getBlockZ() shr 4
    }

    fun getVectorWhenXIsOnLine(other: Vector, x: Float): Vector? {
        val xDiff = other.x - this.x
        val yDiff = other.y - y
        val zDiff = other.z - z
        val f = (x - this.x) / xDiff
        return if (f in 0f..1f) Vector(this.x + xDiff * f, y + yDiff * f, z + zDiff * f, dimension) else null
    }

    fun getVectorWhenYIsOnLine(other: Vector, y: Float): Vector? {
        val xDiff = other.x - x
        val yDiff = other.y - this.y
        val zDiff = other.z - z
        val f = (y - this.y) / yDiff
        return if (f in 0f..1f) Vector(x + xDiff * f, this.y + yDiff * f, z + zDiff * f, dimension) else null
    }

    fun getVectorWhenZIsOnLine(other: Vector, z: Float): Vector? {
        val xDiff = other.x - x
        val yDiff = other.y - y
        val zDiff = other.z - this.z
        val f = (z - this.z) / zDiff
        return if (f in 0f..1f) Vector(x + xDiff * f, y + yDiff * f, this.z + zDiff * f, dimension) else null
    }

    open fun add(x: Float, y: Float, z: Float): Vector {
        return Vector(this.x + x, this.y + y, this.z + z, dimension)
    }

    open fun subtract(x: Float, y: Float, z: Float): Vector {
        return Vector(this.x - x, this.y - y, this.z - z, dimension)
    }

    open fun multiply(x: Float, y: Float, z: Float): Vector {
        return Vector(this.x * x, this.y * y, this.z * z, dimension)
    }

    open fun divide(x: Float, y: Float, z: Float): Vector {
        return Vector(this.x / x, this.y / y, this.z / z, dimension)
    }

    fun normalize(): Vector {
        val squaredLength = squaredLength()
        return if (squaredLength < 1.0E-4) Vector(0f, 0f, 0f) else divide(squaredLength, squaredLength, squaredLength)
    }

    fun distance(vector: Vector): Float {
        return sqrt(distanceSquared(vector).toDouble()).toFloat()
    }

    fun distanceSquared(vector: Vector): Float {
        return (FastMath.pow((x - vector.x).toDouble(), 2) + FastMath.pow(
            (y - vector.y).toDouble(),
            2
        ) + FastMath.pow((z - vector.z).toDouble(), 2)).toFloat()
    }

    fun squaredLength(): Float {
        return FastMath.sqrt((x * x + y * y + z * z).toDouble()).toFloat()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val vector = other as Vector
        return vector.x.compareTo(x) == 0 && vector.y.compareTo(y) == 0 && vector.z.compareTo(z) == 0 && dimension === vector.dimension
    }

    override fun hashCode(): Int {
        return Objects.hash(x, y, z, dimension)
    }

    @SneakyThrows
    public override fun clone(): Vector {
        val clone = super.clone() as Vector
        clone.x = x
        clone.y = y
        clone.z = z
        return clone
    }

    fun getIntermediateWithXValue(var1: Vector, x: Float): Vector? {
        val var4: Float = var1.getX() - this.getX()
        val var6: Float = var1.getY() - this.getY()
        val var8: Float = var1.getZ() - this.getZ()
        return if (var4 * var4 < 1.0000000116860974E-7) {
            null
        } else {
            val var10: Float = (x - this.getX()) / var4
            if (!(var10 < 0.0) && !(var10 > 1.0)) Vector(
                this.getX() + var4 * var10,
                this.getY() + var6 * var10,
                this.getZ() + var8 * var10
            ) else null
        }
    }

    fun getIntermediateWithYValue(var1: Vector, y: Float): Vector? {
        val var4: Float = var1.getX() - this.getX()
        val var6: Float = var1.getY() - this.getY()
        val var8: Float = var1.getZ() - this.getZ()
        return if (var6 * var6 < 1.0000000116860974E-7) {
            null
        } else {
            val var10: Float = (y - this.getY()) / var6
            if (!(var10 < 0.0) && !(var10 > 1.0)) Vector(
                this.getX() + var4 * var10,
                this.getY() + var6 * var10,
                this.getZ() + var8 * var10
            ) else null
        }
    }

    fun getIntermediateWithZValue(var1: Vector, z: Float): Vector? {
        val var4: Float = var1.getX() - this.getX()
        val var6: Float = var1.getY() - this.getY()
        val var8: Float = var1.getZ() - this.getZ()
        return if (var8 * var8 < 1.0000000116860974E-7) {
            null
        } else {
            val var10: Float = (z - this.getZ()) / var8
            if (!(var10 < 0.0) && !(var10 > 1.0)) Vector(
                this.getX() + var4 * var10,
                this.getY() + var6 * var10,
                this.getZ() + var8 * var10
            ) else null
        }
    }

    override fun toString(): String {
        return "Vector{" +
                "x=" + x +
                ", y=" + y +
                ", z=" + z +
                ", dimension=" + dimension +
                '}'
    }

    fun round(): Vector {
        return Vector(x.roundToInt(), y.roundToInt(), z.roundToInt())
    }

    companion object {
        fun zero(): Vector {
            return Vector(0, 0, 0)
        }
    }
}

